### ISSUE TITLE
add import ruins my (ns) declaration

### ISSUE THREAD

**Issue by gdanov** [clojure-emacs/clj-refactor.el#154]
Following the example animated for add import, I trigger the operation, type in my import statement and when I exit the snippet I end up with namespace declaration where the existing ```(:require )``` is gone and I am left with empty ```(:import)``` 

Examples
* before:
```
(ns jj.repl
	(:require [puget.printer]
						[clojure.tools.namespace.repl :as ns-repl]
            ;; [jj.gram.core]
            jj.gram.printer
            jj.gram.state-machine
            [jj.core :as core]
            jj.gram.group-exec
            )
)
```
* after
```
(ns jj.repl
	(:require)
```

Note that it also unbalances the parentheses. Not only of the ns statement but on the subsequent as well.
using latest stable versions

**Comment by gdanov** [comment]
hm, looks like the comment causes the damage

**Comment by benedekfazekas** [comment]
it is still a valid error report.

a workaround tho: run `clean-ns` (`cn`) before adding import

**Comment by expez** [comment]
I'm not going to fix this.  It's not within scope to update `cljr-sort-ns` to support non-standard formatting.  

As an aside, I consider `cljr-sort-ns` to be replaced by `cljr-clean-ns`, but we can't run `cljr-clean-ns` after adding a new import because it removes stuff which isn't used and presumably you haven't started using the newly imported class yet.

tl;dr format your code like this:

```clj
(ns jj.repl
    (:require [puget.printer]
                        [clojure.tools.namespace.repl :as ns-repl]
            ;; [jj.gram.core]
            jj.gram.printer
            jj.gram.state-machine
            [jj.core :as core]
            jj.gram.group-exec))
```

**Comment by magnars** [comment]
Agreed with expez. Use the standard formatting.

**Comment by benedekfazekas** [comment]
seems i was overruled but with a good reason ;)

**Comment by gdanov** [comment]
Sorry, I don't understand what is the non-standard in the formatting. Your suggestion for proper formatting still includes the commented line and as discussed above this causes serious mess up.

**Comment by magnars** [comment]
The problem is the dangling parens:

```
            jj.gram.group-exec
            )
)
```

replacing it with the standard formatting fixes the problem. The comment is ok.

```
            jj.gram.group-exec))
```

**Comment by gdanov** [comment]
thank you that makes it clearer. I am surprised cljr is opinionated about formatting to the point where it treats it as significant as if this was coffeescript and not lisp-y language.

Even if I agree with how the formatting must be done, I cannot agree that this type of corruption is acceptable outcome. It's both surprising and unclear. Doubt you want to create this wtf impression.

Refusing to complete the operation or some other action with feedback how to correct the problem would be much better user experience.

**Comment by expez** [comment]
> I am surprised cljr is opinionated about formatting to the point where it treats whitespace as significant as if this was coffeescript and not lisp-y language.

@gdanov this was part of the motivation for writing `cljr-clean-ns` to replace all the other commands which operate on the ns.  `cljr-clean-ns` relies on the clojure reader itself to support whatever formatting the user chooses.  

You can turn off the use of `cljr-sort-ns` by setting `cljr-auto-sort-ns` to `nil`.  This means you won't get sorting automatically after adding an import, but you can keep whatever formatting you want.

As a friendly aside I'll note that the preference for placing all trailing parenthesis on a single line is so dominant in the lisp community (after being a convention for 60 years) that it's more akin to a [more](http://en.wikipedia.org/wiki/Mores) than an opinion.

**Comment by gdanov** [comment]
please note I am **not**  focused on discussing which formatting is right on wrong, but rather on the behavior overall. Whitespace being insignificant, I see the formatting as orthogonal concern which has no place in this discussion, regardless if we agree or not. 

As stated above, this destructive behavior is hard to sell as expected side effect of the operation, so please reconsider the "won't fix" label.

Thank you for the hint, I will definitely give it a go.

**Comment by expez** [comment]
I definitely agree with you that this is bad, but I won't fix it because:

1. `cljr-sort-ns`, which contains the problem, is going to be deprecated shortly.
2. The fraction of users using this kind of formatting is so small the bug is bordering on irrelevant.
3. Fixing it would be a ton of work.

**Comment by gdanov** [comment]
thanks
